src/gc-object-layout.h

```diff
@@ -0,0 +1,199 @@
+#else
+#define jl_astaggedvalue(v)                                             \
+    ((jl_taggedvalue_t*)((char*)(v) - sizeof(jl_taggedvalue_t)))
+#define jl_valueof(v)                                                   \
+    ((jl_value_t*)((char*)(v) + sizeof(jl_taggedvalue_t)))
+#define jl_typeof(v)                                                    \
+    jl_to_typeof(jl_typetagof(v))
+#define jl_typetagof(v)                                                 \
+    ((jl_astaggedvalue(v)->header) & ~(uintptr_t)15)
+#endif
+static inline void jl_set_typeof(void *v, void *t) JL_NOTSAFEPOINT
+{
+    // Do not call this on a value that is already initialized.
+    jl_taggedvalue_t *tag = jl_astaggedvalue(v);
+    jl_atomic_store_relaxed((_Atomic(jl_value_t*)*)&tag->type, (jl_value_t*)t);
+}
+
+// =======
+// task layout
+// =======
+
+// all values are callable as Functions
+typedef struct _jl_value_t jl_function_t;
+struct _jl_tls_states_t;
+
+#define JL_RNG_SIZE 5 // xoshiro 4 + splitmix 1
+
+typedef struct _jl_task_t {
+    JL_DATA_TYPE
+    struct _jl_value_t *next; // invasive linked list for scheduler
+    struct _jl_value_t *queue; // invasive linked list for scheduler
+    struct _jl_value_t *tls;
+    struct _jl_value_t *donenotify;
+    struct _jl_value_t *result;
+    struct _jl_value_t *scope;
+    jl_function_t *start;
+    // 4 byte padding on 32-bit systems
+    // uint32_t padding0;
+    uint64_t rngState[JL_RNG_SIZE];
+    _Atomic(uint8_t) _state;
+    uint8_t sticky; // record whether this Task can be migrated to a new thread
+    _Atomic(uint8_t) _isexception; // set if `result` is an exception to throw or that we exited with
+    // 1 byte padding
+    // uint8_t padding1;
+    // multiqueue priority
+    uint16_t priority;
+
+// hidden state:
+
+    // id of owning thread - does not need to be defined until the task runs
+    _Atomic(int16_t) tid;
+    // threadpool id
+    int8_t threadpoolid;
+    // Reentrancy bits
+    // Bit 0: 1 if we are currently running inference/codegen
+    // Bit 1-2: 0-3 counter of how many times we've reentered inference
+    // Bit 3: 1 if we are writing the image and inference is illegal
+    uint8_t reentrant_timing;
+    // 2 bytes of padding on 32-bit, 6 bytes on 64-bit
+    // uint16_t padding2_32;
+    // uint48_t padding2_64;
+    // saved gc stack top for context switches
+    struct _jl_gcframe_t *gcstack;
+    size_t world_age;
+    // quick lookup for current ptls
+    struct _jl_tls_states_t *ptls; // == jl_all_tls_states[tid]
+#ifdef USE_TRACY
+    const char *name;
+#endif
+    // saved exception stack
+    struct _jl_excstack_t *excstack;
+    // current exception handler
+    struct _jl_handler_t *eh;
+    // saved thread state
+    jl_ucontext_t ctx;
+    void *stkbuf; // malloc'd memory (either copybuf or stack)
+    size_t bufsz; // actual sizeof stkbuf
+    unsigned int copy_stack:31; // sizeof stack for copybuf
+    unsigned int started:1;
+} jl_task_t;
```

Comment:
I don't think the task structure should be here. It is not part of GC or have anything to do with object layout.

Reply:
Agreed.